src/metrics/include/metrics/operation.hpp

```diff
@@ -0,0 +1,260 @@
+    using EventSeries = TimeSeries<ClockSource, OperationEvent<ClockSource>>;
+
+    OperationImpl(std::string actorName, std::string opName, EventSeries& events)
+        : _actorName(std::move(actorName)),
+          _opName(std::move(opName)),
+          _events(std::addressof(events)) {}
+
+    /**
+     * @return the name of the actor running the operation.
+     */
+    const std::string& getActorName() const {
+        return _actorName;
+    }
+
+    /**
+     * @return the name of the operation being run.
+     */
+    const std::string& getOpName() const {
+        return _opName;
+    }
+
+    void reportAt(time_point finished, OperationEvent<ClockSource>&& event) {
+        _events->addAt(finished, event);
+    }
+
+private:
+    const std::string _actorName;
+    const std::string _opName;
+    EventSeries* const _events;
+};
+
+/**
+ * RAII type for managing data captured about a running operation. Constructing an instance starts a
+ * timer for how long the operation takes. The timer ends when success() or failure() is called on
+ * the instance.
+ */
+template <typename ClockSource>
+class OperationContextT : private boost::noncopyable {
+private:
+    using OutcomeType = typename OperationEvent<ClockSource>::OutcomeType;
+
+public:
+    using time_point = typename ClockSource::time_point;
+
+    explicit OperationContextT(v1::OperationImpl<ClockSource>& op)
+        : _op{std::addressof(op)}, _started{ClockSource::now()} {}
+
+    OperationContextT(OperationContextT<ClockSource>&& other) noexcept
+        : _op{std::move(other._op)},
+          _started{std::move(other._started)},
+          _event{std::move(other._event)},
+          _isClosed{std::exchange(other._isClosed, true)} {}
+
+    ~OperationContextT() {
+        if (!_isClosed) {
+            BOOST_LOG_TRIVIAL(warning)
+                << "Metrics not reported because operation '" << _op->getOpName()
+                << "' being run by actor '" << _op->getActorName()
+                << "' did not close with success() or failure().";
+        }
+    }
+
+    /**
+     * Increments the counter for the number of iterations. This method only needs to be called if
+     * an actor is periodically reporting its operations. By default an `iters = 1` value is
+     * automatically reported.
+     */
+    void addIterations(count_type iters) {
+        _event.iters += iters;
+    }
+
+    /**
+     * Increments the counter for the number of documents inserted, modified, deleted, etc.
+     */
+    void addDocuments(count_type ops) {
+        _event.ops += ops;
+    }
+
+    /**
+     * Increments the size in bytes of the number of documents inserted, etc.
+     */
+    void addBytes(count_type size) {
+        _event.size += size;
+    }
+
+    /**
+     * Increments the counter for the number of write errors, transient transaction errors, etc.
+     */
+    void addErrors(count_type errors) {
+        _event.errors += errors;
+    }
```

Comment:
maybe default the params to 1 for all the `add*` methods?

Reply:
I'm not sold on more syntax sugar helping all that much. I think not needing to call `addIterations(1)` altogether is the biggest win.